### PR TITLE
Command module with centralized Command Pattern logic

### LINKED ISSUE #983: Stop using system.exit for fluo commands
See https://github.com/apache/fluo/pull/982/files/a4bf630e6de398a848d412f6ca1a11bfb4dee6fa#r158413399

### PR DESCRIPTION
**Pull request by jkosh44**
This is my first pass attempt at a POC for using the command pattern to implement the Fluo command module. It still needs some polishing up and I want to add tests but I wanted to give people a chance to review and comment before continuing on. It doesn't conform to the command pattern 100% but it's pretty close, and I found forcing it to conform 100% was either making it too verbose or unnecessarily complicated.

This ended up touching just about every file in the command module and became a decently big change. 

I didn't attempt to solve the System.out/System.err vs Logger problem as I described in Issue #983  but it may be something to consider while reviewing.

The overall goal of this PR is to make it easier to reuse the commands outside of the command line (in a UI or API for example) while moving the logic to run them from the command line to a central location.

Fixes #983

### CODE REVIEW

File: modules/command/src/main/java/org/apache/fluo/command/FluoCommand.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.fluo.command;
+
+@FunctionalInterface
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/FluoCommand.java]: I thought it may be nice to be able to run a command as a lambda sometime in the future, however I don't actually take advantage of this annotation. It may be unnecessary and without it I could remove the default implementation for `isHelp()`.

File: modules/command/src/main/java/org/apache/fluo/command/FluoCommandRunner.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.fluo.command;
+
+public class FluoCommandRunner {
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/FluoCommandRunner.java]: This class could probably be combined with the `FluoProgram` class. It might be better to have the logic of both of these classes in the same place. I'm interested to hear others opinions on it.

From the Command Pattern I tried to make this the Invoker class.

File: modules/command/src/main/java/org/apache/fluo/command/BaseCommand.java [hunk 3]
```diff
@@ -15,29 +15,14 @@
 
 package org.apache.fluo.command;
 
-import com.beust.jcommander.JCommander;
 import com.beust.jcommander.Parameter;
-import com.beust.jcommander.ParameterException;
-
-public class BaseOpts {
 
+public abstract class BaseCommand implements FluoCommand {
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/BaseCommand.java]: I went back and forth on whether or not I should keep the options in the same class as the functionality for each command, or separate them into different classes. I ultimately decided that it would be more concise to have them be in the same class. Though I'm interested in other peoples opinions.

File: modules/command/src/main/java/org/apache/fluo/command/FluoExec.java [hunk 4]
```diff
@@ -44,25 +50,39 @@ protected void configure() {
     }
   }
 
-  public static void main(String[] args) throws Exception {
-    if (args.length < 2) {
-      System.err.println("Usage: fluo exec <app> <class> args...");
-      System.exit(-1);
+  @Parameter(description = "<app> <class> args...", variableArity = true)
+  private List<String> args;
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/FluoExec.java]: It would be nice to be able to have three fields `String app`, `String class`, and `List<String> args`. However I couldn't get this to work with JCommander which is why there's just a single field `List<String> args`. I'm going to keep trying to see if I can get it to work with three fields.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import static org.apache.fluo.command.FluoCommandType.INIT;
+import static org.apache.fluo.command.FluoCommandType.LIST;
+import static org.apache.fluo.command.FluoCommandType.ORACLE;
+import static org.apache.fluo.command.FluoCommandType.REMOVE;
+import static org.apache.fluo.command.FluoCommandType.SCAN;
+import static org.apache.fluo.command.FluoCommandType.STATUS;
+import static org.apache.fluo.command.FluoCommandType.WAIT;
+import static org.apache.fluo.command.FluoCommandType.WORKER;
+
+public class FluoProgram {
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: I'm now leveraging the command functionality form JCommander, which we weren't previously leveraging. You can read about it here [http://jcommander.org/#_more_complex_syntaxes_commands](http://jcommander.org/#_more_complex_syntaxes_commands).

From the Command Pattern I tried to make this class the Client class.

File: modules/distribution/src/main/scripts/fluo [hunk 6]
```diff
@@ -131,50 +131,50 @@ function setup_service {
 case "$1" in
```
Comment (jkosh44) [inline modules/distribution/src/main/scripts/fluo]: I'm not as proficient at shell scripting as I'd like to be so any comments on this file would be greatly appreciated. Now that the same class is being used as the entry point for almost all commands this script looks a little redundant. I feel like there may be ways of condensing it, but I didn't want to attempt to much in this PR.

File: modules/command/src/main/java/org/apache/fluo/command/FluoCommandRunner.java (hunk 2 above)
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoCommandRunner.java]: I am in favor of moving it to FluoProgram.  If you inlined the static method runFluoCommand then all calls to System.exit may be in FluoProgram.main.

File: modules/command/src/main/java/org/apache/fluo/command/BaseCommand.java (hunk 3 above)
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/BaseCommand.java]: I like the options in the same class.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        return;
+    }
+
+    String programName = String.format("fluo %s", parsedCommandType.getCommandName());
+    JCommander command = jcommand.findCommandByAlias(parsedCommandType.getCommandName());
+    command.setProgramName(programName);
+
+    if (parsedCommand.isHelp()) {
+      command.usage();
+      System.exit(1);
```
Comment (keith-turner) [inline modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: Why exit(1) when someone runs help?

File: modules/distribution/src/main/scripts/fluo (hunk 6 above)
Comment (keith-turner) [reply modules/distribution/src/main/scripts/fluo]: It does look like it could be condensed, but I am not sure of the best approach.  Would be a good follow on issue.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      parsedCommandOrProgram.setProgramName(String.format("fluo %s", commandName));
+      parsedCommandOrProgram.usage();
+      System.exit(1);
+      return;
+    }
+
+    FluoCommandType parsedCommandType = FluoCommandType.valueOf(jcommand.getParsedCommand());
+
+    FluoCommand parsedCommand;
+    switch (parsedCommandType) {
```
Comment (keith-turner) [inline modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: Looking at the Jcommander API I was wondering if it would be possible to replace this switch statement with `Jcommander.getObjects()`

File: modules/command/src/main/java/org/apache/fluo/command/FluoConfig.java [hunk 9]
```diff
@@ -17,22 +17,28 @@
...
     try (FluoAdmin admin = FluoFactory.newAdmin(config)) {
       for (Map.Entry<String, String> entry : admin.getApplicationConfig().toMap().entrySet()) {
         System.out.println(entry.getKey() + " = " + entry.getValue());
       }
     }
   }
+
+  @Override
+  public void execute() throws FluoCommandException {
+    config();
```
Comment (keith-turner) [inline modules/command/src/main/java/org/apache/fluo/command/FluoConfig.java]: Could the be inlined?

File: modules/command/src/main/java/org/apache/fluo/command/FluoExec.java (hunk 4 above)
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoExec.java]: I saw a subparameter annotation that has an order parameter in the jcommander docs.   Maybe that could be used to accomplish this, not sure.

File: modules/command/src/main/java/org/apache/fluo/command/FluoExec.java [hunk 10]
```diff
@@ -44,25 +50,39 @@ protected void configure() {
...
+
+      // inject fluo configuration
+      Guice.createInjector(new FluoConfigModule(clazz, fluoConfig));
 
-    // inject fluo configuration
-    Guice.createInjector(new FluoConfigModule(clazz, fluoConfig));
+      Method method = clazz.getMethod("main", String[].class);
+      method.invoke(null, args.subList(2, args.size()));
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+      throw new FluoCommandException(String.format("Class %s must have a main method", className),
```
Comment (keith-turner) [inline modules/command/src/main/java/org/apache/fluo/command/FluoExec.java]: Nice error messages.

File: modules/command/src/main/java/org/apache/fluo/command/FluoGetJars.java [hunk 11]
```diff
@@ -81,4 +70,9 @@ public static void main(String[] args) {
       log.error("", e);
     }
   }
+
+  @Override
+  public void execute() throws FluoCommandException {
+    getJars();
```
Comment (keith-turner) [inline modules/command/src/main/java/org/apache/fluo/command/FluoGetJars.java]: Could this be inlined?

File: modules/command/src/main/java/org/apache/fluo/command/FluoList.java [hunk 12]
```diff
@@ -18,18 +18,18 @@
...
+@Parameters(commandDescription = "Lists all Fluo applications in Fluo instance")
+public class FluoList extends ConfigCommand {
 
-  public static void main(String[] args) throws Exception {
-
-    ConfigOpts commandOpts = ConfigOpts.parse("fluo list", args);
+  public void list() {
     FluoConfiguration config = CommandUtil.resolveFluoConfig();
-    commandOpts.overrideFluoConfig(config);
+    overrideFluoConfig(config);
```
Comment (keith-turner) [inline modules/command/src/main/java/org/apache/fluo/command/FluoList.java]: Seeing this pattern repeated I was wondering if it would make sense for the super class to have a method like.

```java
FluoConfiguration getConfig() {
  FluoConfiguration config = CommandUtil.resolveFluoConfig();
  overrideFluoConfig(config);
  return config.
}
```

Not sure if this is good idea because I have not looked into what `resolveFluoConfig()` does.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+    try {
+      jcommand.parse(args);
+    } catch (ParameterException e) {
+      System.err.println(e.getMessage());
+      String commandName = Optional.ofNullable(jcommand.getParsedCommand()).orElse("");
+      JCommander parsedCommandOrProgram =
+          Optional.ofNullable(jcommand.findCommandByAlias(commandName)).orElse(jcommand);
+      parsedCommandOrProgram.setProgramName(String.format("fluo %s", commandName));
+      parsedCommandOrProgram.usage();
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: Yes. If I run the following program
```
FluoConfig fluoConfig = new FluoConfig();
    FluoExec fluoExec = new FluoExec();
    FluoGetJars fluoGetJars = new FluoGetJars();
    FluoInit fluoInit = new FluoInit();
    FluoList fluoList = new FluoList();
    FluoOracle fluoOracle = new FluoOracle();
    FluoRemove fluoRemove = new FluoRemove();
    FluoScan fluoScan = new FluoScan();
    FluoStatus fluoStatus = new FluoStatus();
    FluoWait fluoWait = new FluoWait();
    FluoWorker fluoWorker = new FluoWorker();
    JCommander jcommand = JCommander.newBuilder().addCommand(CONFIG.getCommandName(), fluoConfig)
        .addCommand(EXEC.getCommandName(), fluoExec)
        .addCommand(GET_JARS.getCommandName(), fluoGetJars)
        .addCommand(INIT.getCommandName(), fluoInit).addCommand(LIST.getCommandName(), fluoList)
        .addCommand(ORACLE.getCommandName(), fluoOracle)
        .addCommand(REMOVE.getCommandName(), fluoRemove).addCommand(SCAN.getCommandName(), fluoScan)
        .addCommand(STATUS.getCommandName(), fluoStatus).addCommand(WAIT.getCommandName(), fluoWait)
        .addCommand(WORKER.getCommandName(), fluoWorker).build();

    try{
      jcommand.parse("asdf", "-3dd");
    } catch (ParameterException e) {
      System.err.println(e.getMessage());
      String commandName = Optional.ofNullable(jcommand.getParsedCommand()).orElse("");
      JCommander parsedCommandOrProgram =
          Optional.ofNullable(jcommand.findCommandByAlias(commandName)).orElse(jcommand);
      parsedCommandOrProgram.setProgramName(String.format("fluo %s", commandName));
      parsedCommandOrProgram.usage();
      System.exit(1);
      return;
    }
```

Because it's unable to parse a command then I will see the following in my console
```
Expected a command, got asdf
Usage: fluo  [command] [command options]
  Commands:
    config      Prints application configuration stored in Zookeeper for <app>
      Usage: config [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    exec      Executes <class> with <args> using classpath for <app>
      Usage: exec <app> <class> args...

    get-jars      Copies <app> jars from DFS to local <dir>
      Usage: get-jars [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
        * -d
            Download directory path
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    init      Initializes Fluo application for <app> using <appProps>
      Usage: init [options]
        Options:
          --clearTable
            Skips prompt and clears Accumulo table
            Default: false
          --clearZookeeper
            Skips prompt and clears Zookeeper
            Default: false
          -f, --force
            Skip all prompts and clears Zookeeper and Accumulo table.  
            Equivalent to setting both --clearTable --clearZookeeper
            Default: false
          -h, -help, --help
            Prints help
          --retrieveProperty
            Gets specified property without initializing
          -u, --update
            Update Fluo configuration in Zookeeper
            Default: false
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []
        * -p
            Path to application properties file

    list      Lists all Fluo applications in Fluo instance
      Usage: list [options]
        Options:
          -h, -help, --help
            Prints help
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    oracle      Starts Fluo Oracle process for <app>
      Usage: oracle [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    remove      Removes Fluo application for <app>
      Usage: remove [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    scan      Prints snapshot of data in Fluo <app>
      Usage: scan [options]
        Options:
          -esc, --escape-non-ascii
            Hex encode non ascii bytes
          -h, -help, --help
            Prints help
          --json
            Export key/values stored in Accumulo as JSON file.
          --ntfy
            Scan active notifications
          --raw
            Show underlying key/values stored in Accumulo. Interprets the data 
            using Fluo internal schema, making it easier to comprehend.
        * -a
            Fluo application name
          -c
            Columns of scan in comma separated format: 
            <<columnfamily>[:<columnqualifier>]{,<columnfamily>[:<columnqualifier>]}> 
          -e
            End row (inclusive) of scan
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []
          -p
            Row prefix to scan
          -r
            Exact row to scan
          -s
            Start row (inclusive) of scan

    status      Prints status of Fluo application for <app>
      Usage: status [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    wait      Waits until all notifications are processed for <app>
      Usage: wait [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []

    worker      Starts Fluo Worker process for <app>
      Usage: worker [options]
        Options:
          -h, -help, --help
            Prints help
        * -a
            Fluo application name
          -o
            Override configuration set in properties file. Expected format: -o 
            <key>=<value> 
            Default: []
```
If I change `"asdf", "-3dd"` to `"list", "-3dd"` then because it can parse out a command the following is printed to my console
```
Was passed main parameter '-3dd' but no main parameter was defined in your arg class
Usage: fluo list [options]
  Options:
    -h, -help, --help
      Prints help
    -o
      Override configuration set in properties file. Expected format: -o 
      <key>=<value> 
      Default: []
```
The usage outputs are coming from the descriptions from the `@Parameter` and `@Paramters` annotations.

File: modules/command/src/main/java/org/apache/fluo/command/FluoExec.java (hunk 4 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoExec.java]: I tried using `SubParameter` first actually, but the combination of having a main parameter with sub-parameters on a command seemed to confuse JCommander. I'll try playing around with it some more.

File: modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java [hunk 14]
```diff
@@ -17,18 +17,12 @@
 
 import com.beust.jcommander.Parameter;
 
-class CommonOpts extends ConfigOpts {
+abstract class CommonCommand extends ConfigCommand {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java]: I think that name is fine. I chose `CommandCommand` to keep it consistent with `CommonOpts`, but I think `AppCommand` may be more descriptive.

On a somewhat related note I was thinking of combing `ConfigCommand` and `BaseCommand` into the `BaseCommand` class because I'm not sure if there's any benefit from separating them out.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java (hunk 7 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: I had `System.exit(1)` because that was the existing behavior from `BaseOpts` class. Though now that I'm looking at it we should probably just return here.

File: modules/command/src/main/java/org/apache/fluo/command/FluoList.java (hunk 12 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoList.java]: I'll take a look and see if this makes sense.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java (hunk 8 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: I saw that method as well, but I'm not exactly sure how to take advantage of it. My understanding is that `getObjects()` will return a list of the command objects that were passed in via `addCommand`. But we'll still need to select the correct object depending on which command was parsed.

Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: I was wondering about it behavior in the following case.

```java
JCommander command = jcommand.findCommandByAlias(parsedCommandType.getCommandName());
var objs = command.getObjects(); //does this return all objs or a single obj related to the specific sub command
```

File: modules/command/src/main/java/org/apache/fluo/command/FluoCommand.java (hunk 1 above)
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoCommand.java]: Personally I would remove it if its not used and causes an unnecessary default method.

File: modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java]: Is Config command the only thing that extends BaseCommand?  If so it would make sense to collapse them.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: > If I change "asdf", "-3dd" to "list", "-3dd" then because

That is very cool.

File: modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java]: Yes, `ConfigCommand` is the only thing that extends `BaseCommand`. I'll collapse them.

[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/CommonCommand.java]: Actually I think FluoExec should extend BaseCommand so that you can pass a -h option. So I'll leave them separate.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java (hunk 8 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: That's a good idea. So that will return a `List<Object>` of size 1, with the object being the one related to the specific sub command.
We can replace the switch statement with something like this
```
JCommander command = jcommand.findCommandByAlias(parsedCommandType.getCommandName());
List<Objects> objs = command.getObjects();
if(objs.size() ==1 && objs.get(0) instanceof FluoCommand) {
  FluoCommand parsedCommand = (FluoCommand)objs.get(0);
  ...
} else {
  //Error Handling
}
```

I think that's definitely more concise than the switch statement, and we may even be able to get rid of the enum completely. It does feel slightly less clean though with the size and instance of check.

File: modules/command/src/main/java/org/apache/fluo/command/FluoList.java (hunk 12 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoList.java]: So it seems like the pattern for resolving the config is as follows
First `CommandUtil.resolveFluoConfig()` is called which gets the config from the file defined by the fluo.conn.props System property.
Then just the `FluoInit` class lets you override configs through a file with the -p option
Then if the command extends AppCommand it will let you override the application name via the -a option
Then it will override all the configs that you provide via the -o option.

The only exception is that FluoExec is structured differently than the other commands, so I'll leave it out of this specific discussion.

I think it would make sense to give `ConfigCommand` the following method
```
FluoConfiguration getConfig() {
    FluoConfiguration config = CommandUtil.resolveFluoConfig();
    overrideFluoConfig(config);
    return config;
  }
```
and five `AppCommand` the following method
```
@Override
  FluoConfiguration getConfig() {
    FluoConfiguration config = super.getConfig();
    config.setApplicationName(applicationName);
    return config;
  }
```

The only issue with this is that the order of overriding configurations may change. This would only be an issue if someone is overriding configurations in different places with different values. For example if someone gave an application name via the -a option but then gave a different application name via the -o option. I don't think this is really a valid use case so we should be fine.

File: modules/command/src/main/java/org/apache/fluo/command/FluoExec.java [hunk 15]
```diff
@@ -44,25 +50,39 @@ protected void configure() {
     }
   }
 
-  public static void main(String[] args) throws Exception {
-    if (args.length < 2) {
-      System.err.println("Usage: fluo exec <app> <class> args...");
-      System.exit(-1);
+  @Parameter(description = "<app> <class> args...", variableArity = true)
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/FluoExec.java]: Another quick note about this class, unlike the other commands it parses out the application name by position not by the -a option. Also it does not have a -o option for overriding configuration. It makes it slightly harder to reuse functionality and slightly breaks the consistency of the other commands. May not be something we address in this PR, but just wanted to throw that out there.

File: modules/command/src/main/java/org/apache/fluo/command/FluoList.java [hunk 16]
```diff
@@ -54,17 +54,37 @@ public static void main(String[] args) throws Exception {
...
+        listApp(config, path);
+      }
+    }
+  }
+
+  private boolean checkCuratorExists(CuratorFramework curator) {
+    try {
+      return curator.checkExists().forPath("/") != null;
+    } catch (Exception e) {
+      throw new RuntimeException("Error checking if curator exists", e);
```
Comment (jkosh44) [inline modules/command/src/main/java/org/apache/fluo/command/FluoList.java]: `.forPath` throws Exception and I thought that it was a case were we'd want to show the whole stack trace. I don't think this isn't the cleanest way, so I'm open to suggestions.

File: modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java (hunk 8 above)
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoProgram.java]: It the expectation is that its always a FluoCommand and of size one, then the following code would throw an exception when this is not the case w/o the if stmt checking size and instanceof.

```java
  FluoCommand parsedCommand = (FluoCommand)Iterables.getOnlyElement(command.getObjects());
```

File: modules/command/src/main/java/org/apache/fluo/command/FluoList.java (hunk 16 above)
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/FluoList.java]: this seems ok to me.  Could add a comment saying why this was done.   Also its possible that Exception could be a RuntimeException.  Could do.

```java
try {
￼      return curator.checkExists().forPath("/") != null;
   } catch (RuntimeException e) {
      throw e; 
￼    } catch (Exception e) {
      // want stack trace for this exception to show up on command line
￼      throw new RuntimeException("Error checking if curator exists", e);
   }
```

File: modules/command/src/main/java/org/apache/fluo/command/FluoExec.java (hunk 4 above)
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/FluoExec.java]: Been experimenting a little more, and I briefly looked through the source code. I don't think it's possible.

**Comment by jkosh44** [comment]
So I'm pretty happy with the current status of this PR and I think I've addressed all the comments. If I've missed any or if there are other issues please let me know. I'm going to work on adding tests and doing some functional testing. It might take a little while because I need to figure out the best way to test all the commands.

There's a couple of potential follow up issues though that I've come across

1. Investigate possible ways of condensing the [fluo script](https://github.com/apache/fluo/pull/1083/files#diff-9085d61b756f1527aae51ea144142951).
After this change the [fluo script](https://github.com/apache/fluo/pull/1083/files#diff-9085d61b756f1527aae51ea144142951) has become pretty repetitive, there may be some ways of condensing it.
2. Figure out solution for printing to logger vs `System.out`. 
Some commands print to `System.out` while others log to a Logger
3. Make commands more friendly to be used outside of CLI.
(This may be very related to number 2). Because `execute` returns `void`, the caller doesn't get much information from the commands, just whether it was successful or if there was an Exception. For example `FluoList` is pretty useless outside of the CLI right now, it just prints the running applications to `System.out`. Issues like #670 could benefit from being able to use the info from `FluoList`. One possible solution would be to move the functionality to another method like `public T list()` where `T` was some type that had the relevant information for listing applications. Then `execute()` could call `list()` and print the information to `System.out`. The web app mentioned in #670  could just call `list()` directly. The other commands would each have to be looked at individually.

**Comment by keith-turner** [comment]
> So I'm pretty happy with the current status of this PR and I think I've addressed all the comments.

I think this is in really good shape.  I like the recent change making command names an annotation parameter.

> Investigate possible ways of condensing the fluo script.

Could possibly open a follow on issue and advertise that on Apache helpwanted as a bash task. I like narrow and very specific issues for helpwanted, and this feels like one.

> Figure out solution for printing to logger vs System.out.

Could open up a follow on issue for that.

**Comment by jkosh44** [comment]
Thanks for the help reviewing @keith-turner. I'm hoping to finish up the tests sometime after the holidays, probably in the new year. I can open up those issues after a final review and merge. I like the idea of putting the bash issue on help wanted. The other issue I'll probably just grab myself unless someone else is particularly interested.

**Comment by keith-turner** [comment]
@jkosh44 what testing have you done for this? I plan to pull this PR down locally and try running it.  Before I did that I was curious what testing you had done.  Also, are there any significant outstanding changes you still plan to make for this PR?  If so, I will hold of on testing it locally.

**Comment by jkosh44** [comment]
@keith-turner after last night's change I have no more planned changes for the actual functionality. I do still want to try and add Integration Tests. I'm not sure how I'm going to do that right now or if it's possible so I plan on looking into that next. 

In terms of testing I used uno and stresso to compare running the commands of the previous version to running the commands on my version. There's still probably some edge cases and optional parameters to test but I'm going to do that after integration tests.

So feel free to pull this down and test it yourself. I don't plan on changing any of the main code unless an issue is found.

**Comment by jkosh44** [comment]
@keith-turner after last night's change I have no more planned changes for the actual functionality. I do still want to try and add Integration Tests. I'm not sure how I'm going to do that right now or if it's possible so I plan on looking into that next.

In terms of testing I used uno and stresso to compare running the commands of the previous version to running the commands on my version. There's still probably some edge cases and optional parameters to test but I'm going to do that after integration tests.

So feel free to pull this down and test it yourself. I don't plan on changing any of the main code unless an issue is found.

I also want to squash all the commits once it's fully complete.

**Comment by keith-turner** [comment]
> I also want to squash all the commits once it's fully complete.

That is something  usually do after review is complete and its approved.  I usually use the Squash and Merge button on github (there is a drop down arrow beside the Create a Merge Commit button) when there are no merge conflcts.  Be careful with the Squash and merge button though.  In the past it failed for me and when it did a retry button showed up.  When I pressed retry it merged all of my commits instead of doing squash and merge.  This has happened to me twice, so if I see the it in the future I am not going to retry.

**Comment by jkosh44** [comment]
@keith-turner Unless you think we should do integration tests as a follow up PR since this has grown fairly big already. In that case we can just do functional testing, squash, final review, and merge.

**Comment by keith-turner** [comment]
@jkosh44 I think integration test as a follow on issue is a good way to go.  I will run this locally today or tomorrow and after that I will be finished reviewing.

File: modules/command/src/main/java/org/apache/fluo/command/AppCommand.java [hunk 17]
```diff
@@ -16,19 +16,25 @@
...
+
   String getApplicationName() {
     return applicationName;
   }
 
-  public static CommonOpts parse(String programName, String[] args) {
-    CommonOpts opts = new CommonOpts();
-    parse(programName, opts, args);
-    return opts;
+  void setApplicationName(String applicationName) {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/AppCommand.java]: I meant to standardize how all of the options are handled but I forgot to. I wanted them all to be private fields with a public getter and setter method. The reason for this is because if the commands are going to be used outside of the CLI, I wanted to provide a way to set the options without having to use Reflection. 

This reply applies for your comment on `setProperties` too.

What are you thoughts?

File: modules/command/src/test/java/org/apache/fluo/command/FluoProgramTest.java [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    }
+  }
+
+  private MockFluoCommand mockFluoCommand;
+
+  @BeforeClass
+  public static void disablePrinting() {
+    // This will hide usage and error logs when running tests
+    try (PrintStream ps = new PrintStream(new NullOutputStream())) {
+      System.setOut(ps);
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/test/java/org/apache/fluo/command/FluoProgramTest.java]: Done

File: modules/command/src/test/java/org/apache/fluo/command/ConfigCommandTest.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+
+    assertEquals(1, nullSplitList.size());
+    assertEquals(testStr, nullSplitList.get(0));
+  }
+
+  @Test
+  public void testGetConfig() {
+    int newZookeeperTimeout = 100;
+    List<String> overrideConfig = Collections.singletonList(
+        FluoConfiguration.CONNECTION_ZOOKEEPER_TIMEOUT_PROP + "=" + newZookeeperTimeout);
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/test/java/org/apache/fluo/command/ConfigCommandTest.java]: Done

File: modules/command/src/main/java/org/apache/fluo/command/AppCommand.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply modules/command/src/main/java/org/apache/fluo/command/AppCommand.java]: > The reason for this is because if the commands are going to be used outside of the CLI

Do you have a specific use case for that in mind?

> What are you thoughts?

This internal non-API code so we can change it at anytime, which is nice.  We can commit this now, release Fluo 2.0.0, and then change this w/o any problem.  

There is a balancing act between making what you are committing now more understandable for others and committing things in a partially done state because you know you will work on it in the future.  Whatever you think is best I am ok with.

File: modules/command/src/test/java/org/apache/fluo/command/ConfigCommandTest.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply modules/command/src/test/java/org/apache/fluo/command/ConfigCommandTest.java]: Nice

File: modules/command/src/main/java/org/apache/fluo/command/AppCommand.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (jkosh44) [reply modules/command/src/main/java/org/apache/fluo/command/AppCommand.java]: The use case I have in mind is if we ever want to wrap the commands in some type of UI as mentioned in the original issue.

I think I'll make your suggested changes and we should be good to merge. I can figure out how to set the options in a follow up PR.

**Comment by jkosh44** [comment]
@keith-turner Alright all the comments have been addressed. Is it ok to merge my own PR or should I let someone else do it?
